### ISSUE TITLE
Swift binding does not handle polling error

### ISSUE THREAD

**Issue by conanoc** [mozilla/uniffi-rs#1514]
This is part of generated Swift code:
```Swift
private func _UniFFI_AskarStore_Close_waker(raw_env: UnsafeMutableRawPointer?) {
    Task {
        let env = Unmanaged<_UniFFI_AskarStore_Close_Env>.fromOpaque(raw_env!)
        let env_ref = env.takeUnretainedValue()
        let polledResult = UnsafeMutableRawPointer.allocate(byteCount: 0, alignment: 0)
        do {
            let isReady = try rustCallWithError(FfiConverterTypeErrorCode.self) {
                uniffi_aries_askar_fn_method_askarstore_close_poll(
                    env_ref.rustFuture,
                    _UniFFI_AskarStore_Close_waker,
                    env.toOpaque(),
                    polledResult,
                    $0
                )
            }

            if isReady {
                env_ref.continuation.resume(returning: ())
                polledResult.deallocate()
                env.release()
            }
        } catch {
            env_ref.continuation.resume(throwing: error)
            polledResult.deallocate()
            env.release()
        }
    }
}
```

There is no `else` condition for `isReady`.  `AskarStore.Close()` will hang forever when `isReady` is false. I think it should resume with error when `isReady` is false.

Another strange thing is that `uniffi_aries_askar_fn_method_askarstore_close_poll` is taking `_UniFFI_AskarStore_Close_waker` function as a waker parameter. What's the purpose of the waker parameter?

Below is the generated scaffolding code:
```Rust
#[doc(hidden)]
#[no_mangle]
pub extern "C" fn uniffi_aries_askar_fn_method_askarstore_close(
    this: <::std::sync::Arc<AskarStore> as ::uniffi::FfiConverter<crate::UniFfiTag>>::FfiType,
    call_status: &mut ::uniffi::RustCallStatus,
) -> Box<::uniffi::RustFuture<Result<(), ErrorCode>>> {
    ::uniffi::deps::log::debug!("close");
    Box::new(::uniffi::RustFuture::new_tokio(async move {
        <::std::sync::Arc<AskarStore> as ::uniffi::FfiConverter<crate::UniFfiTag>>::try_lift(this)
            .unwrap_or_else(|err| ::std::panic!("Failed to convert arg 'self': {}", err))
            .close()
            .await
    }))
}
#[doc(hidden)]
#[no_mangle]
pub extern "C" fn uniffi_aries_askar_fn_method_askarstore_close_poll(
    future: ::std::option::Option<&mut ::uniffi::RustFuture<Result<(), ErrorCode>>>,
    waker: ::std::option::Option<::uniffi::RustFutureForeignWakerFunction>,
    waker_environment: *const ::std::ffi::c_void,
    polled_result: &mut ::std::mem::MaybeUninit<
        <Result<(), ErrorCode> as ::uniffi::FfiConverter<crate::UniFfiTag>>::ReturnType,
    >,
    call_status: &mut ::uniffi::RustCallStatus,
) -> bool {
    ::uniffi::ffi::uniffi_rustfuture_poll::<_, crate::UniFfiTag>(
        future,
        waker,
        waker_environment,
        polled_result,
        call_status,
    )
}
```

**Comment by conanoc** [comment]
I tested it again to check the `isReady` value and it was `false` only for `AskarStore.close()` function. In the case of other async functions, `true` was returned always.
Is there a way to find the reason?

**Comment by bendk** [comment]
> There is no else condition for isReady. AskarStore.Close() will hang forever when isReady is false. I think it should resume with error when isReady is false.

> Another strange thing is that uniffi_aries_askar_fn_method_askarstore_close_poll is taking _UniFFI_AskarStore_Close_waker function as a waker parameter. What's the purpose of the waker parameter?

These two things go together.  It's expected that `isReady` is sometimes false, in which case the poll shouldn't do anything.  In that case the waker will be called sometime later when the future might be ready, which will schedule a poll, and the process should repeat again.

Can you post a minimal example showing a future that always hangs?

**Comment by conanoc** [comment]
I see. `_UniFFI_AskarStore_Close_waker` can be called several times and the final `isReady` value will be `true`.
In my case, `_UniFFI_AskarStore_Close_waker` is called only once with `false`. This means that the Rust function for `AskarStore.close()` may be blocked indefinitely for some reason.
Let me check first if the Rust function works when it is called in a Rust program.

**Comment by conanoc** [comment]
I've tested the `AskarStore.close()` function in Rust world and it worked without any problem.

There is one suspicious part, but I'm not sure how to test it.
- RustFuture is using `async_compat` to create a tokio runtime. `async_compat` [doc](https://docs.rs/async-compat/latest/async_compat/) says it creates a "context of a global single-threaded tokio runtime" for a `Future`. The code is here: https://docs.rs/async-compat/latest/src/async_compat/lib.rs.html#428
- `AskarStore.close()` closes sqlite connection pools with `sqlx`. I found an issue saying that pool close would hang if it is run on single-threaded runtime. https://github.com/launchbadge/sqlx/issues/1928#issuecomment-1218781267

I would appreciate any advice.

**Comment by conanoc** [comment]
I'm still struggling to create a simplified example to reproduce this issue. I created an example that should reproduce this issue but it is producing a different issue for now.
https://github.com/conanoc/uniffi_test

This example crashes while creating a table on sqlite store.

![crash](https://user-images.githubusercontent.com/11764352/234744339-83a5181f-a30c-4126-9054-241795e16312.jpg)

**Comment by conanoc** [comment]
The crash was due to the fact that `_UniFFI_CreateStore_waker()` was called several times before `uniffi_bug_finder_fn_func_create_store_poll()` returns. The waker function was called even after, actually in between, the poll returned true.
I could prevent the crash using a lock like this:
```Swift
let lock = NSLock()
var jobFinished = false
private func _UniFFI_CreateStore_waker(raw_env: UnsafeMutableRawPointer?) {
    print("_UniFFI_CreateStore_waker called")
    Task {
        lock.withLock {
            if jobFinished {
                print("job finished. skip polling.")
                return
            }
            let env = Unmanaged<_UniFFI_CreateStore_Env>.fromOpaque(raw_env!)
            let env_ref = env.takeUnretainedValue()
            let polledResult = UnsafeMutablePointer <UnsafeMutableRawPointer>.allocate(capacity: 1)
            
            print("poll >>>>")
            let isReady = try! rustCall {
                uniffi_bug_finder_fn_func_create_store_poll(
                    env_ref.rustFuture,
                    _UniFFI_CreateStore_waker,
                    env.toOpaque(),
                    polledResult,
                    $0
                )
            }
            
            print("<<<< got poll result: \(isReady)")
            if isReady {
                env_ref.continuation.resume(returning: try! FfiConverterTypeStore.lift(polledResult.move()))
                polledResult.deallocate()
                env.release()
                jobFinished = true
            }
        }
    }
}
```

This is what the test prints:
```
Test Case '-[BugFinderTests.BugFinderTests testStore]' started.
testStore starts
_UniFFI_CreateStore_waker called
poll >>>>
Database already exists
<<<< got poll result: false
_UniFFI_CreateStore_waker called
poll >>>>
<<<< got poll result: false
_UniFFI_CreateStore_waker called
poll >>>>
<<<< got poll result: false
_UniFFI_CreateStore_waker called
poll >>>>
<<<< got poll result: false
_UniFFI_CreateStore_waker called
poll >>>>
_UniFFI_CreateStore_waker called
_UniFFI_CreateStore_waker called
_UniFFI_CreateStore_waker called
_UniFFI_CreateStore_waker called
Create table result: SqliteQueryResult { changes: 0, last_insert_rowid: 0 }
<<<< got poll result: true
job finished. skip polling.
job finished. skip polling.
job finished. skip polling.
job finished. skip polling.
In count
count: 0
In close
before close
after close
store closed
Test Case '-[BugFinderTests.BugFinderTests testStore]' passed (1.022 seconds).
```

Unfortunately, however, the phenomenon of the close function being hung has not been reproduced.

**Comment by bendk** [comment]
I'm not sure, but I think you may be running into https://github.com/mozilla/uniffi-rs/issues/1507, or a related issue.  I just merged a large refactor of the async code that was meant to fix this.  Can you test against the current main and see if helps?

**Comment by conanoc** [comment]
@bendk After rebuild with `cargo update`, the crashes are gone. Thanks.

I changed my sample a little bit to reproduce my original issue. [here](https://github.com/conanoc/uniffi_test/commit/48f8a89b55e1051be93d675f11e2f746c372958c)
This fails to compile and it seems to tell something about the original issue. The error message is this:
```console
error[E0277]: `NonNull<libsqlite3_sys::bindings::sqlite3>` cannot be shared between threads safely
   --> /Users/conanoc/work/rust/bug-finder/target/debug/build/bug-finder-c9063819d7e755e5/out/bug_finder.uniffi.rs:130:1
    |
130 | #[::uniffi::ffi_converter_interface(tag = crate::UniFfiTag)]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `NonNull<libsqlite3_sys::bindings::sqlite3>` cannot be shared between threads safely
    |
    = help: within `Session`, the trait `Sync` is not implemented for `NonNull<libsqlite3_sys::bindings::sqlite3>`
    = note: required because it appears within the type `sqlite::connection::handle::ConnectionHandleRaw`
    = note: required because it appears within the type `sqlite::connection::worker::ConnectionWorker`
    = note: required because it appears within the type `SqliteConnection`
    = note: required because it appears within the type `pool::connection::Live<Sqlite>`
    = note: required because it appears within the type `Option<pool::connection::Live<Sqlite>>`
    = note: required because it appears within the type `PoolConnection<Sqlite>`
note: required because it appears within the type `Session`
   --> src/lib.rs:19:12
    |
19  | pub struct Session {
    |            ^^^^^^^
note: required by a bound in `Interface`
   --> /Users/conanoc/.cargo/git/checkouts/uniffi-rs-6f89edd2a1ffa4bd/58e0a24/uniffi_core/src/lib.rs:262:33
    |
262 | pub trait Interface<UT>: Send + Sync + Sized {
    |                                 ^^^^ required by this bound in `Interface`
    = note: this error originates in the attribute macro `::uniffi::ffi_converter_interface` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `bug-finder` due to previous error
```

**Comment by conanoc** [comment]
@bendk Thanks. I wrapped the connection with `tokio::sync::Mutex` and succeeded in compiling it.
And finally, I could reproduce `Store.close()` function hang.
Could you take a look at this example? https://github.com/conanoc/uniffi_test

**Comment by conanoc** [comment]
I've found the cause of this issue. `Store.close()` hangs because the `PoolConnection` of sqlx is not released after drop.
Why it's not released? Because it needs a tokio runtime.

`sqlx` releases `PoolConnection` on drop like this:
https://github.com/launchbadge/sqlx/blob/v0.5.13/sqlx-core/src/pool/connection.rs#L145
```Rust
/// Returns the connection to the [`Pool`][crate::pool::Pool] it was checked-out from.
impl<DB: Database> Drop for PoolConnection<DB> {
    fn drop(&mut self) {
        if self.live.is_some() {
            #[cfg(not(feature = "_rt-async-std"))]
            if let Ok(handle) = sqlx_rt::Handle::try_current() {
                handle.spawn(self.return_to_pool());
            }

            #[cfg(feature = "_rt-async-std")]
            sqlx_rt::spawn(self.return_to_pool());
        }
    }
}
```

`Session` holds `PoolConnection`, so  `PoolConnection` is dropped when `Session` is dropped. `Session` is dropped automatically when the variable is out of scope in Swift, and the Rust counterpart proceeds the drop without tokio runtime.
A workaround would be to implement an async `close` function in `Session` and call it manually. What do you think? @bendk

**Comment by mhammond** [comment]
> A workaround would be to implement an async close function in Session and call it manually.

While that will probably work for you, I don't think we would be happy to have that as a formal workaround to code unexpectedly hanging as it is being dropped.

> Session is dropped automatically when the variable is out of scope in Swift, and the Rust counterpart proceeds the drop without tokio runtime.

I'm afraid I don't know much about either swift or async rust 😅 but your example explicitly implements `Drop` for `Session` - when you say "the Rust counterpart proceeds the drop without tokio runtime", what is the "Rust counterpart" if not that explicit drop implementation? Why can't that explicit drop reference the tokio runtime?

**Comment by bendk** [comment]
> While that will probably work for you, I don't think we would be happy to have that as a formal workaround to code unexpectedly hanging as it is being dropped.

I think this is really the key issue here.  If this is a workaround for a bug in async_compat or tokio, then it would be okay.  If this is a workaround for the fact that UniFFI is doing the wrong thing, then we should fix our async code.

In particular, I've always been very confused about how UniFFI should interact with tokio or any other Rust async runtime.  Is there something we should be doing better to integrate with others?

**Comment by conanoc** [comment]
I think this is a bug of sqlx. Since the drop() function in the Drop trait is not asynchronous, it is advisable for sqlx not to invoke async functions within it.
It would be helpful if there were an async version of `drop()` in the Drop trait. I found this link: https://rust-lang.github.io/async-fundamentals-initiative/roadmap/async_drop.html

**Comment by conanoc** [comment]
This change resolved the issue: https://github.com/conanoc/uniffi_test/commit/0b746d39bb441bbdef9e240f9ed6b44186a4aa3d
I created an async function `Session.close()` and dropped the `PoolConnection` there.
